### ISSUE TITLE
win_pkg: 2017.7 Branch Issues

### ISSUE THREAD

**Issue by damon-atkins** [saltstack/salt#43908]
### Description of Issue/Question
The following are issues with the existing 2017.7 (assumed to be also in 2017.7.2)

"New Results" - refers to the latest 2016.11 #43708 branch (note this version of win_pkg is dependant on the cmdmod in 2016.11 branch with recent windows fix)
"Original Results" - refers to 2017.7 branch win_pkg.py

pkg.list_upgrades
```
New Results DIFFERENT  32                                    Original Results 29
[INFO    ] win_pkg.refresh_db Removing all *.sls files       [INFO    ] win_pkg.refresh_db Removing all *.sls files 
under                                                        under 
'c:\salt\var\cache\salt\minion\files\base\win\repo-ng'       'c:\salt\var\cache\salt\minion\files\base\win\repo-ng'
[INFO    ] fileclient.cache_dir Caching directory            [INFO    ] fileclient.cache_dir Caching directory 
'win/repo-ng/' for environment 'base'                        'win/repo-ng/' for environment 'base'
                                                           > [ERROR   ] __init__.version_cmp LooseVersion instance has 
                                                           > no attribute 'version'
                                                           > Traceback (most recent call last):
                                                           >   File 
                                                           > "C:\salt\bin\lib\site-packages\salt\utils\__init__.py", 
                                                           > line 2584, in version_cmp
                                                           >     if _LooseVersion(pkg1) < _LooseVersion(pkg2):
                                                           >   File "C:\salt\bin\lib\distutils\version.py", line 296, 
                                                           > in __cmp__
                                                           >     return cmp(self.version, other.version)
                                                           > AttributeError: LooseVersion instance has no attribute 
                                                           > 'version'
                                                           ------- cut ----- repeated error msg -------
                                                           > [ERROR   ] __init__.version_cmp LooseVersion instance has 
                                                           > no attribute 'version'
                                                           > Traceback (most recent call last):
                                                           >   File 
                                                           > "C:\salt\bin\lib\site-packages\salt\utils\__init__.py", 
                                                           > line 2584, in version_cmp
                                                           >     if _LooseVersion(pkg1) < _LooseVersion(pkg2):
                                                           >   File "C:\salt\bin\lib\distutils\version.py", line 296, 
                                                           > in __cmp__
                                                           >     return cmp(self.version, other.version)
                                                           > AttributeError: LooseVersion instance has no attribute 
                                                           > 'version'
{                                                            {
    "local": {                                                   "local": {
        "chrome": "latest",                                          "chrome": "latest", 
        "nsis": "3.0b2",                                             "nsis": "3.0b2", 
        "git": "2.13.1.2",                                           "git": "2.13.1.2", 
        "adobeshockwaveplayer": "latest",                            "adobeshockwaveplayer": "latest", 
        "grepwin": "1.6.682"                                         "grepwin": "1.6.682"
    }                                                            }
}                                                            }
```

pkg.upgrade_available grepwin saltenv=does_not_exist
saltenv setting ignored.
```
New Results DIFFERENT   4                                    Original Results 33
{                                                            {
    "local": false                                         |     "local": true
}                                                            }
```

pkg.remove grepwin version=1.6.646
Does not handle download msi for an uninstall.
```
New Results DIFFERENT   5                                    Original Results 5
[WARNING ] __init__.warn_until                               [WARNING ] __init__.warn_until 
C:\salt\bin\lib\site-packages\salt\modules\win_update.py:9   C:\salt\bin\lib\site-packages\salt\modules\win_update.py:9
1: DeprecationWarning: The 'win_update' module is being      1: DeprecationWarning: The 'win_update' module is being 
deprecated and will be removed in Salt Fluorine              deprecated and will be removed in Salt Fluorine 
(Unreleased). Please use the 'win_wua' module instead.       (Unreleased). Please use the 'win_wua' module instead.

[INFO    ] cmdmod._run Executing command                   | [INFO    ] cmdmod._run Executing command [u'msiexec', 
'"C:\Windows\system32\cmd.exe" /s /c ""msiexec" /X         | u'/x', '/qn', '/norestart'] in directory 'C:\Users\root'
"c:\salt\var\cache\salt\minion\extrn_files\base\netassist. | [ERROR   ] cmdmod.run_all Command '[u'msiexec', u'/x', 
dl.sourceforge.net\project\grepwin\1.6.14\grepWin-1.6.14-x | '/qn', '/norestart']' failed with return code: 1619
64.msi" /qn /norestart"' in directory 'C:\Users\root'      | [ERROR   ] cmdmod.run_all stdout: This 
                                                           > installation package could 
                                                           > not be opened.  Verify that 
                                                           > the package exists and that 
                                                           > you can access it or contact 
                                                           > the application vendor to 
                                                           > verify that this is a valid 
                                                           > Windows Installer package.
                                                           > 
                                                             
                                                           > [ERROR   ] cmdmod.run_all retcode: 1619
                                                           > [ERROR   ] win_pkg.remove Failed to remove grepwin
                                                           > [ERROR   ] win_pkg.remove retcode 1619
                                                           > [ERROR   ] win_pkg.remove uninstaller output: This 
                                                           > installation package could 
                                                           > not be opened.  Verify that 
                                                           > the package exists and that 
                                                           > you can access it or contact 
                                                           > the application vendor to 
                                                           > verify that this is a valid 
                                                           > Windows Installer package.
                                                           > 
                                                             
{                                                            {
    "local": {                                                   "local": {
        "grepwin": {                                                 "grepwin": {
            "new": "",                                     |             "uninstall status": "failed"
            "old": "1.6.673"                               <
        }                                                            }
    }                                                            }
}                                                            }
```

**Comment by gtmanfred** [comment]
@twangboy can you take a look at this when you get a minute?

Thanks,
Daniel

**Comment by damon-atkins** [comment]
Salt team need to decide if they wish the fixes #43708 to be include in the 2017.7.2 release after testing them.

**Comment by twangboy** [comment]
@damon-atkins @gtmanfred So the question is if we want to backport the develop `win_pkg` to 2017.7 like you did with #43708 ....

**Comment by twangboy** [comment]
That would probably be merged forward...

**Comment by damon-atkins** [comment]
Yes?  Just want to point out the issues before 2017.7.2 is released as its branched already.

**Comment by twangboy** [comment]
This wouldn't make it in until 2017.7.3

**Comment by damon-atkins** [comment]
That sounds about right. Given the PR #43708  still waiting on someone to merge it into 2016.11
Might be worth adding something into the release notes for 2017.7.X that `pkg.remove` issues with download msi for an uninstall.  And `pkg.upgrade_available` ignores saltenv.

**Comment by damon-atkins** [comment]
Time to close this. Fixed 2016.11 and its been merged forward.
